### PR TITLE
WF-753: Implement Routing for communication between Journeys

### PR DESCRIPTION
**Pull request by manisha-backbase**
Added make transfer redo feature to pass the matrix query params for passing the data from transaction list journey to make transfer journey.
This can be used as an example for communication between journeys and also a use case for matrix query parameters

### CODE REVIEW

**Comment by manisha-backbase** [comment]
Screenshot 2022-09-16 at 14 14 23
Screenshot 2022-09-16 at 14 14 16

File: libs/transfer/src/lib/components/make-transfer-form/make-transfer-form.component.ts [hunk 1]
```diff
@@ -65,9 +67,14 @@ export class MakeTransferFormComponent implements OnInit {
     return !!control && control.touched && control.invalid;
   }
 
-  constructor(private fb: FormBuilder) {}
+  constructor(private fb: FormBuilder, private route: ActivatedRoute) {}
 
   ngOnInit(): void {
+    this.route.paramMap.subscribe((params) => {
```
Comment (kirillsud) [inline libs/transfer/src/lib/components/make-transfer-form/make-transfer-form.component.ts:73]: Missing `unsubscribe`

File: apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts [hunk 2]
```diff
@@ -30,6 +33,13 @@ export class AchPositivePayInterceptor implements HttpInterceptor {
       );
     }
 
-    return next.handle(req);
+    return next.handle(req).pipe(
+      catchError((err: HttpErrorResponse) => {
+        if (err.status.toString().startsWith('4')) {
+          this.router.navigateByUrl('/error-page');
```
Comment (kirillsud) [inline apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts:39]: What are we demonstrating here?

**Review by kirillsud — CHANGES_REQUESTED** [review]

File: apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts (hunk 2 above)
Comment (manisha-backbase) [reply apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts:39]: There was a use case for handling HTTP errors with routing .

Comment (Farzannajipour) [reply apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts:39]: What will happen if for any reason we get 5xx errors?

File: libs/transactions/src/lib/components/transaction-item/transaction-item.component.html [hunk 3]
```diff
@@ -35,3 +35,19 @@
...
+  <button
+    bbButton
+    [color]="'primary'"
+    [buttonSize]="'sm'"
+    [routerLink]="[
+      '/transfer/make-transfer',
+      {
+        amount: transaction.transactionAmountCurrency.amount,
+        accountName: transaction.merchant?.name || transaction.counterPartyName
+      }
```
Comment (kirillsud) [inline libs/transactions/src/lib/components/transaction-item/transaction-item.component.html:48]: Can we put these parameters between `transfer` and `make-transfer` to add more sense to use them?

File: apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts (hunk 2 above)
Comment (kirillsud) [reply apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts:39]: Maybe also we can pass a custom message in extra params to demonstrate how to use `state` option: https://angular.io/api/router/NavigationBehaviorOptions#state and also we can try `skipLocationChange` https://angular.io/api/router/NavigationBehaviorOptions#skipLocationChange

File: libs/transfer/src/lib/components/make-transfer-form/make-transfer-form.component.ts [hunk 4]
```diff
@@ -65,9 +67,14 @@ export class MakeTransferFormComponent implements OnInit {
     return !!control && control.touched && control.invalid;
   }
 
-  constructor(private fb: FormBuilder) {}
+  constructor(private fb: FormBuilder, private route: ActivatedRoute) {}
```
[the remark this answers is not in the document]
Comment (manisha-backbase) [reply libs/transfer/src/lib/components/make-transfer-form/make-transfer-form.component.ts:70]: Done

File: apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts (hunk 2 above)
Comment (manisha-backbase) [reply apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts:40]: As discussed, I have removed this. First, We should understand how Journey's implement error handling. We can have a separate task for this.

File: libs/transactions/src/lib/components/transaction-item/transaction-item.component.html (hunk 3 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/components/transaction-item/transaction-item.component.html:48]: Attached a sceenshot

**Comment by manisha-backbase** [comment]
Screenshot 2022-09-21 at 20 55 46

File: libs/transactions/src/lib/components/transaction-item/transaction-item.component.html (hunk 3 above)
Comment (kirillsud) [reply libs/transactions/src/lib/components/transaction-item/transaction-item.component.html:52]: Oh noo, you can do it by using `snapshot.parent`

Comment (kirillsud) [reply libs/transactions/src/lib/components/transaction-item/transaction-item.component.html:52]: But your option is also nice

File: libs/transfer/src/lib/transfer-journey.component.html [hunk 5]
```diff
@@ -1,3 +1,12 @@
+<div class="row" *ngIf="counterPartyName">
+  <bb-alert-ui
+    title="Transfer Alert"
+    modifier="info"
+    [dismissible]="true"
+    [message]="'Making Repeated Transfer for ' + counterPartyName"
```
Comment (kirillsud) [inline libs/transfer/src/lib/transfer-journey.component.html:6]: Can we add translations for that?

Comment (kirillsud) [reply libs/transfer/src/lib/transfer-journey.component.html:6]: And for the title?

**Review by kirillsud — CHANGES_REQUESTED** [review]

File: libs/transactions/src/lib/components/transaction-item/transaction-item.component.html (hunk 3 above)
Comment (manisha-backbase) [reply libs/transactions/src/lib/components/transaction-item/transaction-item.component.html:52]: But is it a good practice to use snapshot.parent in child components. If something needs to be used in the entire route hierarchy one should go for queryparams.

File: apps/golden-sample-app/src/app/app-routing.module.ts [hunk 6]
```diff
@@ -18,6 +18,11 @@ const routes: Routes = [
       ),
     canActivate: [AuthGuard],
   },
+  {
+    path: 'error-page',
+    loadChildren: () =>
+      import('./error-page/error-page.module').then((m) => m.ErrorPageModule),
+  },
```
[the remark this answers is not in the document]
Comment (manisha-backbase) [reply apps/golden-sample-app/src/app/app-routing.module.ts:25]: Can we keep it for future use in the demo app. As its a standalone module for error?

File: apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts [hunk 7]
```diff
@@ -8,9 +8,11 @@ import {
 } from '@angular/common/http';
 import { Observable, timer } from 'rxjs';
 import { map } from 'rxjs/operators';
+import { Router } from '@angular/router';
```
[the remark this answers is not in the document]
Comment (manisha-backbase) [reply apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts:11]: Done

File: apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts [hunk 8]
```diff
@@ -8,9 +8,11 @@ import {
...
 
 @Injectable()
 export class AchPositivePayInterceptor implements HttpInterceptor {
+  constructor(private router: Router) {}
```
[the remark this answers is not in the document]
Comment (manisha-backbase) [reply apps/golden-sample-app/src/app/interceptors/ach-positive-pay.interceptor.ts]: Done

File: libs/transfer/src/lib/transfer-journey.component.html (hunk 5 above)
Comment (manisha-backbase) [reply libs/transfer/src/lib/transfer-journey.component.html:6]: I have created a separate issue for this to add translations.

File: apps/golden-sample-app/src/app/app-routing.module.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (kirillsud) [reply apps/golden-sample-app/src/app/app-routing.module.ts:25]: Move it to another branch/PR then because we do not sure about it is a way we recommend or not

**Comment by santam85** [comment]
Are the comments all addressed? Is this ready to review again?

File: libs/transactions/src/lib/components/transaction-item/transaction-item.component.html (hunk 3 above)
Comment (kirillsud) [reply libs/transactions/src/lib/components/transaction-item/transaction-item.component.html:52]: Yes. It's alright if the parent route is a part of the journey

**Comment by kirillsud** [comment]
I've resolved all comments. Below you can find some descriptions of my changes.

### Communication between journeys

There are two optional parameters for the route `/transfer;amount=800;accountName=Pocket%20Transfer/..`. `accountName` is used by all children routes to keep the info message. This example demonstrates how that parameters could be passed in the middle of the route and could be optional.

image

### Localisation

1. Added an npm script `start:nl` to start a local web server with NL locale
2. Added a query parameter for the identity login URL to pass the current locale
3. Added translations for new UI elements

### UI improvments

1. UI was improved by reusing the `bb-product-item-basic-account-ui` component to list user accounts
2. Grid improvements for all views

image

**Review by kirillsud — APPROVED** [review]

File: apps/golden-sample-app/src/app/guards/auth.guard.ts [hunk 9]
```diff
@@ -1,16 +1,19 @@
...
+    private oAuthService: OAuthService,
+    @Inject(LOCALE_ID) private locale: string
+  ) {}
 
   canActivate(): boolean {
     if (this.oAuthService.hasValidAccessToken()) {
       return true;
     }
-    this.oAuthService.initLoginFlow();
+    this.oAuthService.initLoginFlow(undefined, { ui_locales: this.locale });
```
Comment (santam85) [inline apps/golden-sample-app/src/app/guards/auth.guard.ts:16]: Can we move this change to a different PR? I'm interested in reviewiing it with the Identity team

File: libs/transactions/src/lib/components/text-filter/text-filter.component.html [hunk 10]
```diff
@@ -2,7 +2,7 @@
   placeholder="filter transactions by type"
   aria-label="filter transactions"
   i18n-aria-label="
-    filter transaction placeholder@@transaction.form.filter.aria-label"
+    filter transaction aira label@@transaction.form.filter.aria-label"
```
Comment (santam85) [inline libs/transactions/src/lib/components/text-filter/text-filter.component.html:5]: - [ ] `aria`?

**Review by santam85 — APPROVED** [review]

File: apps/golden-sample-app/src/app/guards/auth.guard.ts (hunk 9 above)
Comment (kirillsud) [reply apps/golden-sample-app/src/app/guards/auth.guard.ts:16]: Done

**Review by kirillsud — APPROVED** [review]
